### ISSUE TITLE
Most recent directional key priority 'swallows' previous held keys

### ISSUE THREAD

**Issue by LunarLambda** [AltimorTASDK/altinput#1]
In the regular typex_loader version of the game, holding a direction to move a piece, and inputting a second direction (i.e. up to hard drop), will result in the next piece moving immediately

When running with tgm3_input, the next piece doesn't move until the first directional button is released and held again

While this arguably emulates a 4-way stick better, it is highly unexpected.

Is this deliberate? Unfortunately I cannot currently recompile tgm3_input

I walked through the keyboard::update function and to my understanding of the code, this shouldn't be happening at all, since direction_keys stores directions for as long as they're held...

**Comment by AltimorTASDK** [comment]
Yes, it's deliberate. IIRC left/right are prioritized over down when diagonals are allowed. You could add

```
if (!(buttons & mask_down))
	buttons |= dir_buttons & mask_up;
```

To the end of `keyboard::update`/`joystick::update` to make an exception for up.

Why aren't you able to recompile it?

**Comment by LunarLambda** [comment]
I'm running everything through Wine, so I don't have a Windows build environment

I'm confused as to what happens exactly

Holding left should push `mask_left` to `direction_keys` (game receives left input)
Pressing up should push `mask_up` to `direction_keys` (-> up input)
Releasing up should remove `mask_up` (-> left input?)

**Comment by ghost** [comment]
Hi.
I didn't test it, but looking at the code it seems broken to me.
By the end of the 'keyboard::update' function, the 'buttons' member variable has at most one directional bit set.
Because of that the 'dir_buttons' variable can only have directional bits from the new event and maybe one bit from the last event.
Directional inputs that were set prior to that are forgotten and processed as 'released' further down in the code.
The easiest way to fix it and to make it work as intended would probably be to add something like
buttons |= old_dir_buttons;
just before the
https://github.com/AltimorTASDK/altinput/blob/master/tgm3_input/keyboard.cpp#L59
line, to restore the actual state of directional inputs before computing what was pressed or released.

EDIT:
Well, I have just installed MSVC and tested it. The fix seems to be working.
In case anyone wants to try it out before it finds its way to an official build, here is the unofficial binaries I've compiled:
[altinput_fix.zip](https://github.com/AltimorTASDK/altinput/files/5488959/altinput_fix.zip)

**Comment by alex-ong** [comment]
@rkodinets I know its a one liner, but could you make a PR?

**Comment by ghost** [comment]
@alex-ong done.
